frontend/src/metabase-lib/expressions/diagnostics.ts

```diff
@@ -88,19 +90,61 @@ export function diagnose({
       : mismatchedParentheses < -1
       ? t`Expecting ${-mismatchedParentheses} opening parentheses`
       : null;
+
   if (message) {
     return { message };
   }
 
+  // try to compile on FE
+  let mbqlOrError: Expr | ErrorWithMessage;
   try {
-    return prattCompiler({ source, startRule, name, query, stageIndex });
+    mbqlOrError = prattCompiler({ source, startRule, name, query, stageIndex });
+
+    if (isErrorWithMessage(mbqlOrError)) {
+      return mbqlOrError;
+    }
+
+    if (startRule === "expression" && isBooleanExpression(mbqlOrError)) {
+      throw new ResolverError(
+        t`Custom columns do not support boolean expressions`,
+        mbqlOrError.node,
+      );
+    }
   } catch (err) {
+    console.warn("compile error", err);
+
     if (isErrorWithMessage(err)) {
       return err;
     }
 
     return { message: t`Invalid expression` };
   }
+
+  // now diagnose on BE
```

Comment:
Would "BE" not confuse a newcomer? They might assume the browser talks to the BE, but nothing like that is happening. I think eventually we should correct our terminology and talk about ML instead of BE (or something similar).